README.md
# Gradient Progress Indicator


## Introduction

This package shows a circular progress indicator with gradient colors, and it is possible insert texts inside that.

# Basic Usage

```dart
class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: Scaffold(
        backgroundColor: Colors.blue,
        body: GradientProgressIndicator(
          radius: 120,
          duration: 3,
          strokeWidth: 12,
          gradientStops: const [
            0.2,
            0.8,
          ],
          gradientColors: const [
            Colors.white,
            Colors.grey,
          ],
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              Text(
                'Some text',
                style: TextStyle(color: Colors.black, fontSize: 18),
              ),
              Text(
                'Another text',
                style: TextStyle(color: Colors.black, fontSize: 18),
              ),
            ],
          ),
        ),
      ),
    );
  }
}
```

![Gradient progress indicator](https://i.giphy.com/media/Smw3omXqeMKxPAvgzn/giphy.webp)

CHANGELOG.md
## 1.0.0+5
Change app description.
## 1.0.0+3
Fixed when the first animation ends and second animation start
## 1.0.0+2
Update README.md and remove internal scaffold
## 1.0.0+1
Update README.md and add tests
## 1.0.0
First version




test/gradient_circular_progress_indicator_test.dart
import 'package:flutter/material.dart';

import 'package:flutter_test/flutter_test.dart';
import 'package:gradient_progress_indicator/gradient_progress_indicator.dart';

void main() {
  testWidgets('should build without exploding', (tester) async {
    await tester.pumpWidget(
      MaterialApp(
        home: Scaffold(
          body: const GradientProgressIndicator(
            gradientColors: [
              Colors.black,
              Colors.white,
            ],
            gradientStops: [
              0.25,
              0.75,
            ],
            radius: 40,
            strokeWidth: 4,
            child: Text('Loading...'),
          ),
        ),
      ),
    );
    await tester.pump(const Duration(seconds: 4));

    expect(find.byType(GradientProgressIndicator), findsOneWidget);
  });

  testWidgets('should build without exploding without gradient colors',
      (tester) async {
    await tester.pumpWidget(
      MaterialApp(
        home: Scaffold(
          body: const GradientProgressIndicator(
            gradientColors: [],
            gradientStops: [],
            radius: 40,
            strokeWidth: 4,
            child: Text('Loading...'),
          ),
        ),
      ),
    );
    await tester.pump(const Duration(seconds: 4));

    // assert
    expect(find.byType(GradientProgressIndicator), findsOneWidget);
  });

  testWidgets('should build without exploding with background color',
      (tester) async {
    // arrange
    await tester.pumpWidget(
      MaterialApp(
        home: Scaffold(
          body: const GradientProgressIndicator(
            gradientColors: [],
            gradientStops: [],
            radius: 40,
            strokeWidth: 4,
            backgroundColor: Colors.red,
            child: Text('Loading...'),
          ),
        ),
      ),
    );
    await tester.pump(const Duration(seconds: 4));

    expect(find.byType(GradientProgressIndicator), findsOneWidget);
  });
}


lib/widget/gradient_progress_indicator_widget.dart
import 'dart:math';

import 'package:flutter/material.dart';

/// This class show the Gradient Progress Indicator
class GradientProgressIndicator extends StatefulWidget {
  const GradientProgressIndicator({
    required this.child,
    required this.radius,
    required this.strokeWidth,
    required this.gradientStops,
    required this.gradientColors,
    this.backgroundColor = Colors.transparent,
    this.duration = 4,
    this.curveType = Curves.easeInQuad,
  });

  final Widget child;
  final int duration;
  final double radius;
  final double strokeWidth;
  final List<double> gradientStops;
  final List<Color> gradientColors;
  final Color backgroundColor;
  final Curve curveType;

  @override
  _GradientProgressIndicatorState createState() =>
      _GradientProgressIndicatorState();
}

// The class is a state class of GradientProgress
class _GradientProgressIndicatorState extends State<GradientProgressIndicator>
    with TickerProviderStateMixin {
  late AnimationController _animationController;
  late AnimationController _animationRotationController;

  @override
  void initState() {
    _animationController = AnimationController(
      vsync: this,
      duration: Duration(seconds: widget.duration),
    );

    _animationRotationController = AnimationController(
      vsync: this,
      duration: Duration(seconds: widget.duration),
    );

    _animationController.addListener(() {
      setState(() {});
      if (_animationController.isCompleted) {
        _animationRotationController.repeat();
      }
    });
    _animationRotationController.addListener(() => setState(() {}));

    _animationController.forward();

    super.initState();
  }

  @override
  void dispose() {
    _animationController.dispose();
    _animationRotationController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return SizedBox(
      child: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        children: <Widget>[
          SizedBox(
            child: Stack(
              alignment: Alignment.center,
              children: [
                Center(
                  child: AnimatedBuilder(
                    animation: _animationRotationController,
                    builder: (_, child) {
                      return Transform.rotate(
                        angle: _animationRotationController.value * (pi * 2),
                        child: child,
                      );
                    },
                    child: _GradientCircularProgressIndicator(
                      gradientColors: widget.gradientColors,
                      radius: widget.radius,
                      strokeWidth: widget.strokeWidth,
                      backgroundColor: widget.backgroundColor,
                      value: Tween(begin: 0.0, end: 1.0)
                          .animate(
                            CurvedAnimation(
                              parent: _animationController,
                              curve: widget.curveType,
                            ),
                          )
                          .value,
                      gradientStops: widget.gradientStops,
                    ),
                  ),
                ),
                Center(
                  child: widget.child,
                ),
              ],
            ),
          ),
        ],
      ),
    );
  }
}

// This class receive the configurations of gradientcircular
class _GradientCircularProgressIndicator extends StatelessWidget {
  _GradientCircularProgressIndicator({
    this.strokeWidth = 10.0,
    this.strokeRound = false,
    this.backgroundColor = Colors.transparent,
    required this.radius,
    required this.gradientStops,
    required this.gradientColors,
    required this.value,
  });

  final double strokeWidth;
  final bool strokeRound;
  final double value;
  final Color backgroundColor;
  final List<double> gradientStops;
  final double radius;
  final List<Color> gradientColors;

  @override
  Widget build(BuildContext context) {
    return Transform.rotate(
      angle: -pi / 2,
      child: CustomPaint(
        size: Size.fromRadius(radius),
        painter: _GradientCircularProgressPainter(
          strokeWidth: strokeWidth,
          strokeRound: strokeRound,
          backgroundColor: backgroundColor,
          gradientColors: gradientColors,
          value: value,
          radius: radius,
          gradientStops: gradientStops,
        ),
      ),
    );
  }
}

// This class paint is gradientcircular
class _GradientCircularProgressPainter extends CustomPainter {
  _GradientCircularProgressPainter({
    required this.strokeWidth,
    required this.strokeRound,
    required this.value,
    this.backgroundColor = Colors.transparent,
    required this.gradientColors,
    required this.gradientStops,
    this.total = 2 * pi,
    this.radius,
  });

  final double strokeWidth;
  final bool strokeRound;
  final double value;
  final Color backgroundColor;
  final List<Color> gradientColors;
  final List<double> gradientStops;
  final double total;
  final double? radius;

  @override
  void paint(Canvas canvas, Size size) {
    Size paintSize = size;
    if (radius != null) {
      paintSize = Size.fromRadius(radius!);
    }

    double _value = value;
    _value = _value.clamp(.0, 0.97) * total;
    const double _start = 0.1;

    final double _offset = strokeWidth / 2;

    final Rect rect = Offset(_offset, _offset) &
        Size(paintSize.width - strokeWidth, paintSize.height - strokeWidth);

    final paint = Paint()
      ..strokeWidth = strokeWidth
      ..strokeCap = StrokeCap.round
      ..style = PaintingStyle.stroke
      ..isAntiAlias = true;

    if (backgroundColor != Colors.transparent) {
      paint.color = backgroundColor;
      canvas.drawArc(rect, _start, total, false, paint);
    }

    if (_value > 0) {
      paint.shader = SweepGradient(
              colors: gradientColors, endAngle: _value, stops: gradientStops)
          .createShader(rect);

      canvas.drawArc(rect, _start, _value, false, paint);
    }
  }

  @override
  bool shouldRepaint(CustomPainter oldDelegate) {
    return true;
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:gradient_progress_indicator/gradient_progress_indicator.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: Scaffold(
        backgroundColor: Colors.blue,
        body: GradientProgressIndicator(
          radius: 120,
          duration: 3,
          strokeWidth: 12,
          gradientStops: const [
            0.2,
            0.8,
          ],
          gradientColors: const [
            Colors.white,
            Colors.grey,
          ],
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              Text(
                'Some text',
                style: TextStyle(color: Colors.black, fontSize: 18),
              ),
              Text(
                'Another text',
                style: TextStyle(color: Colors.black, fontSize: 18),
              ),
            ],
          ),
        ),
      ),
    );
  }
}
